### ISSUE TITLE
Unsupervised Pretraining for Sequence to Sequence Learning

### ISSUE THREAD

**Issue by howardyclo** [howardyclo/papernotes#7]
### Metadata
Authors: Prajit Ramachandran, Peter J. Liu and Quoc V. Le
Organization: Google Brain
Conference: EMNLP 2017
Link: https://goo.gl/n2cKG9

**Comment by howardyclo** [comment]
### Summary

This paper purposes to initialize the weights of encoder and decoder in neural sequence-to-sequence (Seq2Seq) model with two pre-trained language models, and then fine-tuned with labeled data. **During the fine-tuning phase, they jointly train the Seq2Seq objective with the language modeling (LM) objectives to prevent overfitting**. The experiment results show that their method achieves an improvement of 1.3 BLEU from the previous best  models on both WMT'14 and WMT'15 English->German. Human evaluations on abstractive summarization also shows that their method outperforms a purely supervised learning baseline.

![Figure 1](https://github.com/howardyclo/papernotes/blob/master/images/Unsupervised%20Pretraining%20for%20Sequence%20to%20Sequence%20Learning/Figure%201.png?raw=true)

---

### Fine-tuning with Language Modeling Objective
(Note: This is similar to #3 and #5)

After the Seq2Seq is initialized with pre-trained language models (see Figure 1), the Seq2Seq is then fine-tuned with both the translation objective and the language modeling objective (multi-task learning), to avoid **catastrophic forgetting** (GoodFellow et al. 2013). 

### Other Improvements
1. Residual connections: The residual connections can avoid the problem of feeding the random-permuted vector from the task-specific layer to softmax layer, since the task-specific layer is random initialized at first.
2. Multi-layer attention: Attending on both task-specific layer and pre-trained layer.

![Figure 2](https://github.com/howardyclo/papernotes/blob/master/images/Unsupervised%20Pretraining%20for%20Sequence%20to%20Sequence%20Learning/Figure%202.png?raw=true)

---

### Experiments
- Machine translation dataset
    - Training: WMT14 English->Germain dataset, **4 million** after preprocessing.
    - Tokenized with BPE with 89500 merge operations, giving a vocabulary size around 90000.
    - Validation: Concatenated newstest2012 and newstest2013
    - Testing: newstest2014 and newstest2015
- Monolingual training datasets: The News Crawl English and German corpora, each of which has more than **a billion tokens**.
- The language model is trained in the same fashion as Jozefowicz et al. (2016).
- Hyperparameters:
    - Language model: 1-layer 4096 dimensional  LSTM with hidden size projected down to 1024 units. (Trained for **one week** on 32 Tesla K40 GPUs)
    - Seq2Seq: 3-layer model where 2th and 3th layers each have 1000 hidden units.
    - Adam optimizer with 5e-5 learning rate, 0.8 learning rate decay by every 50K steps after an initial 400k steps.
    - Max gradient norm: 5.
    - Dropout: 0.2 on non-recurrent connections.
    - Beam size: 10

---

### Experiment Results
![Table 1](https://github.com/howardyclo/papernotes/blob/master/images/Unsupervised%20Pretraining%20for%20Sequence%20to%20Sequence%20Learning/Table%201.png?raw=true)

---

### Important Insights on Ablation Study

![Figure 3](https://github.com/howardyclo/papernotes/blob/master/images/Unsupervised%20Pretraining%20for%20Sequence%20to%20Sequence%20Learning/Figure%203.png?raw=true)

- For translation, the main gains come from the pre-training features.
- For summarization, pre-training the encoder gives large improvements.
- "Only pre-train decoder" is better than "Only pre-train encoder".
- Pre-train as much as possible.
- Pre-train the softmax (target embeddings) is important.
- The auxiliary LM objective is a strong regularizer. Pre-training the entire model without LM objective is as bad as using the LM objective without pre-training. 
- Pre-training on a large-scale unlabeled data is essential. **If LM is only pre-trained on parallel data (4 million each side) is as bad as not pre-training at all**.

---

### References
- [An Empirical Investigation of Catastrophic Forgeting in Gradient-Based Neural Networks](https://arxiv.org/pdf/1312.6211.pdf) by Ian GoodFellow et al. (2013)
- [Exploring the Limits of Language Modeling](https://arxiv.org/pdf/1602.02410.pdf) by Jozefowicz et al. (2016)
